homeassistant/components/fatek/__init__.py

```diff
@@ -0,0 +1,301 @@
+
+
+class FatekStateCoordinator(DataUpdateCoordinator[None]):
+    """Klasa do komuniakcji / aktualizacji danych z FATEK'a."""
+
+    def __init__(self, hass, fatek_server_config: FatekServerConfig) -> None:
+        """Utworzenie koordnatora."""
+        self._fatek_server_config: FatekServerConfig = fatek_server_config
+        self._m_entities: dict[int, FatekEntity] = {}
+        self._m_register = bitarray(2001)
+        self._m_register.setall(0)
+
+        super().__init__(
+            hass,
+            LOGGER,
+            name=DOMAIN,
+            update_interval=timedelta(seconds=5),
+            request_refresh_debouncer=Debouncer(
+                hass, LOGGER, cooldown=2.0, immediate=False
+            ),
+        )
+
+    @property
+    def fatek_server_config(self) -> FatekServerConfig:
+        """Return the system descriptor."""
+        return self._fatek_server_config
+
+    @property
+    def unique_id(self) -> str:
+        """Return the system descriptor."""
+        entry: ConfigEntry | None = self.config_entry
+        assert entry
+        if entry.unique_id:
+            return entry.unique_id
+        assert entry.entry_id
+        return entry.entry_id
+
+    def add_m_entity(self, entity) -> bool:
+        """Dodaje encję do tablicy."""
+        self._m_entities[entity.index] = entity
+        return True
+
+    def get_m_register(self, index) -> bool:
+        """Zwraca wartość rejestru M dla podanego indeksu."""
+        return self._m_register[index]
+
+    def get_m_registers(self, start, count) -> bool:
+        """Pobranie rejestrów M."""
+        reg_id = f"M_{start}_{count}"
+        str_url = f"http://10.166.5.245:13001/fatek_get?q={reg_id}"
+        # http://myhome.marcinziarko.com:13001/fatek_get?q=set
+        # LOGGER.info("Info. Fatek: FatekStateCoordinator: get_m_registers: %s", str_url)
+        with urllib.request.urlopen(str_url) as url:
+            data = json.load(url)
+            states = data[0][reg_id]
+            # LOGGER.info("Fatek states: %s", states)
+            i: int = 0
+            j: int
+            index: int = start
+            while i < count:
+                j = i * 2
+                val = states[j : j + 1]
+                if val == "0":
+                    if self._m_register[index] == 1:
+                        self.toogle_m_state(index)
+                else:
+                    if self._m_register[index] == 0:
+                        self.toogle_m_state(index)
+                i += 1
+                index += 1
+        return True
+
+    def toogle_m_state(self, index) -> bool:
+        """Zmiana stanu rejestru M."""
+        LOGGER.info("Fatek: FatekStateCoordinator: toogle_m_state: %s", index)
+        self._m_register[index] = not self._m_register[index]
+        if index in self._m_entities:
+            self._m_entities[index].set_entity_state(self._m_register[index])
+        return True
+
+    def set_m_register(self, index: int, value: bool) -> bool:
```

Comment:
It's not quite clear why `toogle_m_state` just modifies the internal state here, but `set_m_register` also does a remote call...